### ISSUE TITLE
Resource Table treats URLs as 1:1 relationship with content, but actual URL-content relationship is many-to-one

### ISSUE THREAD

**Issue by iCaspar** [wp-media/wp-rocket#4199]
**Describe the bug**

Currently, resources we track in the database `rucss_resources` table assume that any given URL points to a unique resource content. In fact, multiple URLs can refer to the same resource. For example,

`https://example.org/file.css?ver=1` can refer to the same content as `https://example.org/file.css?ver=2`.

As happens in the case of some themes/plugins that change version by timestamp under certain conditions, or generate unique URLs for the same content in order to target elements differently on different pages. These cases lead to (at least) 2 undesirable consequences:

1. Duplicate data is stored taking up database resources on both the clients' website and our SaaS database. (See ref to #4161, below.)
2. Attempting to save content with unique hashes results in collisions in SaaS when trying to warmup identical content with different URLs. (See https://github.com/wp-media/warmup/issues/58)

(This issue is the underlying issue for #4161. PR #4187 is the proximate fix, but because of the relation of URLs to content currently in the DB structure on both the Plugin and SaaS side, we cannot implement that fix without introducing a new `major` bug that will cause any content that is referred to from more than one URL to be ALWAYS treated as unprocessed CSS by the treeshaker and therefore NEVER included in the UsedCSS.)

**To Reproduce**
This bug is not directly reproducible _per se_. Rather it us an underlying structural issue in the Database `rucss_resources` table that prevents us from implementing effective fixes for other issues.

**Expected behavior**
The handling of resources needs a means for mapping multiple URLs to a unique resource content.
My first thought is that a lookup table where URL column is a unique index and the `rucss_resources:hash` column is the foreign key may be in order. An identical lookup (or other mapping) would need to be implemented on the SaaS side as well in order for treeshaker to correctly identify warmed-up resources when they are found with different URLs in the page html.

**Additional context**
An identical lookup (or other mapping) would need to be implemented on the SaaS side as well in order for treeshaker to correctly identify warmed-up resources when they are found with different URLs in the page html. Likely, there will need to be corresponding adjustments in each of the DataLayer, Treeshaker and Warmup microservices.

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

**Comment by iCaspar** [comment]
> a lookup table where URL column is a unique index and the rucss_resources:hash column is the foreign key may be in order

The trick with this is that creating this table will require retrieving resources using a JOIN, which is not directly supported (yet) in BerlinDB.
There are ways around it, a couple of which are suggested in [this thread](https://github.com/berlindb/core/issues/63).

In essence we would need 2 queries:
```
// First step: Do we have this URL in the URL_Hash lookup table?
$urlRow = new LookupQuery( [
    'url' => $url
] );

if (! empty $urlRow) { 
    // we have this URL already, all done
    return false;
}

// Second step: Do we have a hash for this URL in the Resources table?
$hash = new ResoucesQuery->get_item_by( 'hash', $urlRow->hash );

if (! empty( $hash ) ) {
    // we have this content, but we need to save the new URL for it 
    LookupQuery->add_item($url, $hash)
    return false;
}

// If we get to here, this is a new resource: Save the resource and the content (in separate tables/queries)
LookupQuery->add_item($url, $hash);
return ResourceQuery->add_item($content, $hash, ...);
```

**Comment by iCaspar** [comment]
Another option raised in discussion with @mostafa-hisham is storing the URL field as an array of URLs. This has the advantage of implementing a parallel solution on the SaaS side (should we move to MongoDB or start sharding in Maria, for example) where we cannot rely on SQL joins.
Downside, it will be hard/expensive to index/search on this.

**Comment by Mai-Saad** [comment]
@GeekPress Do we need to handle the case of safelist URL containing query string? i.e` /test.css?t1=3`
Currently we arenot handling URLs with query string in URL.

**Comment by GeekPress** [comment]
@Mai-Saad If we safelist `/test.css` it should safelist all `/test.css?x=x` cases

**Comment by vmanthos** [comment]
A side-effect of this is that URLs of resources that are dynamically named, e.g.:
https://www.example.com/wp-content/themes/Divi/style-static-cpt.min.divi-switch-4.0.9.css?ver=1630993532

will cause the clearance of the `rucss_used_css` table every time a new page is cached.

This is because a new version of that file will be discovered, due to `ver` changing every time a page is cached. This will lead to `truncate_used_css()` clearing the table through `rocket_rucss_file_changed`:
https://github.com/wp-media/wp-rocket/blob/207346f55a740a6f24f45a0c21e213e18eb5360e/inc/Engine/Optimization/RUCSS/Admin/Subscriber.php#L82

Also, these files will not be processed by the SaaS. Explanation provided by @iCaspar 
> Since the treeshaker will be always looking for a new resource (with the new ver) that is never in the SaaS cache, so even though the resource is there many times (in both the Site and the SaaS, Treeshaker is seeing this URL to access it for the first time every time. 

**Slack discussion:** https://wp-media.slack.com/archives/C029G1B5HD2/p1631101764096500
**Related ticket:** https://secure.helpscout.net/conversation/1623327380/292704?folderId=2135277

**Comment by iCaspar** [comment]
To follow up on @vmanthos note above:

On the treeshaker (SaaS) side, the planned changes to move to an all-redis caching and allow fetching of not-found external-to-SaaS-cache resources to be fetched at treeshake time _should_ resolve the resource duplication/failure-to-include _from the SaaS side_. So at least resources with dynamic urls (mostly `?ver`-related URL changes) will be included in the UsedCss.

On the Rocket side, the issue will remain that `rocket_rucss_file_changed` will still be triggered when finding these types of resources, and new entries in the resource table will continue to be generated.

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1629783217/294372?folderId=2135277

In this case, the following [YITH WooCommerce Product Add-ons & Extra Options Premium](https://yithemes.com/themes/plugins/yith-woocommerce-product-add-ons/) files weren't removed because of the changing `ver` whenever a page is cached:

![](https://i.vgy.me/M5poqm.jpg)

**Comment by vmanthos** [comment]
Related ticket: https://secure.helpscout.net/conversation/1639397647/296648/

The following file wasn't removed:
```
/testimonial-slider-and-showcase/assets/css/testimonial-slider.css?ver=1632814906
```
**Plugin:** [Testimonials Slider and Showcase](https://wordpress.org/plugins/testimonial-slider-and-showcase/)

**Comment by vmanthos** [comment]
Related ticket: https://secure.helpscout.net/conversation/1642516755/297215?folderId=2135277

The following file wasn't removed:
```
/wp-content/themes/porto-child/style.css?noCache=1632811718&ver=5.8.1
```

Activating the feature made navigating from one page to the next slower, probably because of this:
https://github.com/wp-media/wp-rocket/issues/4199#issuecomment-915201993

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1720253627/313178/

The unremoved files were coming from the [OceanWP](https://oceanwp.org/) theme:
```HTML
/wp-content/themes/oceanwp/assets/css/custom_css.css
/wp-content/themes/oceanwp/assets/css/style.min.css
```

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1770086284/322313/

The file kept on this case are the following: 
```
/wp-content/themes/uncode/library/css/style.css?ver=365364507
/wp-content/themes/uncode/library/css/uncode-icons.css?ver=365364507
/wp-content/themes/uncode/library/css/style-custom.css?ver=365364507
```
This leads to showing the wrong fonts and sizes.

**Comment by NataliaDrause** [comment]
Related: https://secure.helpscout.net/conversation/1802833408/329080/

The file:
`/wp-content/uploads/essential-addons-elementor/cb70d11b8.min.css?ver=1646297681`

**Comment by NataliaDrause** [comment]
Related: https://secure.helpscout.net/conversation/1800351708/328614/

This time it doesn't work with the One CDN.

All files that come from `https://usercontent.one/wp/example.com/` have a query string that is changing on each reload.

Examples:
```
https://usercontent.one/wp/example.com/wp-content/plugins/contact-form-7/includes/css/styles.css?ver=5.5.2?media=1647248258
https://usercontent.one/wp/example.com/wp-content/plugins/recent-tweets-widget/tp_twitter_plugin.css?ver=1.0?media=1647248258
https://usercontent.one/wp/example.com/wp-content/plugins/wp-job-openings/assets/css/general.min.css?ver=3.1.0?media=1647248258
```

**Comment by piotrbak** [comment]
Fixed in 3.11. We don't use the resources table anymore.
